segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+import io.pravega.segmentstore.contracts.SegmentType;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.server.ContainerEventProcessor;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+import io.pravega.segmentstore.server.SegmentContainer;
+import io.pravega.segmentstore.server.SegmentStoreMetrics;
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import lombok.Builder;
+import lombok.Cleanup;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static io.pravega.segmentstore.server.containers.MetadataStore.SegmentInfo.newSegment;
+import static io.pravega.shared.NameUtils.getEventProcessorSegmentName;
+
+/**
+ * Implementation for {@link ContainerEventProcessor}. This class stores a map of {@link ContainerEventProcessor.EventProcessor}
+ * identified by name. Once this component is created, it will instantiate new {@link ContainerEventProcessor.EventProcessor}
+ * objects and report metrics for the existing ones. The actual processing is performed in batches by
+ * {@link ContainerEventProcessor.EventProcessor}s. Each of such processors is in charge of tailing their respective
+ * internal Segments and (safely) invoking their handler functions on a list of read events (of at most maxItemsAtOnce
+ * elements). Upon a successful processing iteration, an {@link ContainerEventProcessor.EventProcessor} truncates the
+ * internal Segments of the registered. If an error occurs, the internal Segment is not truncated and the processing is
+ * attempted again over the same events. Therefore, this class provides at-least-once processing guarantees, but events
+ * could be re-processed in the case of failures while truncating the processor's internal Segment. This is important to
+ * take into account when developing handler functions of {@link ContainerEventProcessor.EventProcessor}s as they should
+ * be idempotent and tolerate re-processing.
+ */
+@Slf4j
+class ContainerEventProcessorImpl implements ContainerEventProcessor {
+
+    private static final Duration SHUTDOWN_TIMEOUT = Duration.ofSeconds(10);
+    // Ensure that an Event Processor's Segment does not get throttled by making it system-critical.
+    private final static SegmentType SYSTEM_CRITICAL_SEGMENT = SegmentType.builder().system().internal().critical().build();
```

Comment:
Nit: please be consistent in your modifier order. Above you have `static final` yet here you have `final static`.

Reply:
Done.